Invalidate :has() with nested positional pseudo-class selectors correctly

heycam opened the pull request:
#### 630e5323e6dcbd3c95adb5b8caedb9d8038b445b
<pre>
Invalidate :has() with nested positional pseudo-class selectors correctly
<a href="https://bugs.webkit.org/show_bug.cgi?id=253943">https://bugs.webkit.org/show_bug.cgi?id=253943</a>
rdar://106768224

Reviewed by Antti Koivisto.

ChildChangeInvalidation currently handles :has() invalidation by
traversing the newly inserted or removed subtree and checking whether
there are any :has() rules whose nested selector could match the elements
in the subtree. This is not sufficient to invalidate rules like:

:has(:first-child)
:has(:nth-child(2n+1))
:has(:nth-child(n2+1 of .foo))

When the parent of the changed subtree has any of these flags:

ChildrenAffectedByFirstChildRules
ChildrenAffectedByLastChildRules
ChildrenAffectedByForwardPositionalRules
ChildrenAffectedByBackwardPositionalRules

we must invalidate some siblings of the subtree, to determine if the
:has() selector might have changed.

The &quot;check the existing sibling&quot; optimization that
ChildChangeInvalidation::invalidateForChangedElement does is not valid
if it&apos;s called with a sibling of the changed subtree, so we pass in
a ChangedElementRelation value to control whether to skip it.

* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/child-indexed-pseudo-classes-in-has-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/typed-child-indexed-pseudo-classes-in-has-expected.txt:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::matchHasPseudoClass const):
* Source/WebCore/style/ChildChangeInvalidation.cpp:
(WebCore::Style::ChildChangeInvalidation::invalidateForChangedElement):
(WebCore::Style::ChildChangeInvalidation::invalidateForHasBeforeMutation):
(WebCore::Style::ChildChangeInvalidation::invalidateForHasAfterMutation):
(WebCore::Style::ChildChangeInvalidation::traverseRemovedElements):
(WebCore::Style::ChildChangeInvalidation::traverseAddedElements):
(WebCore::Style::ChildChangeInvalidation::traverseRemainingExistingSiblings):
* Source/WebCore/style/ChildChangeInvalidation.h:

Canonical link: <a href="https://commits.webkit.org/267771@main">https://commits.webkit.org/267771@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0387482caef1269ed66ac5110e5ed34e3b45c655

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17662 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17989 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18521 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19481 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16520 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17856 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21277 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18136 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18602 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17874 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18170 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/15336 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20332 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15404 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16081 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/22667 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16414 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16250 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20525 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16826 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/14250 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15942 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4202 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20307 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16669 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/style/ChildChangeInvalidation.cpp:
```diff
@@ -112,28 +112,98 @@ void ChildChangeInvalidation::invalidateForHasBeforeMutation()
+        invalidateForChangedElement(changedElement, matchingHasSelectors, ChangedElementRelation::SelfOrDescendant);
     });
+
+    auto firstChildStateWillStopMatching = [&](Element* elementAfterChange) {
```
I think elementAfterChange / elementBeforeChange are a bit confusing (given before/after change can refer to the mutation), can we name them `elementNextSibling` / `elementPreviousSibling` or something?

anttijk commented on Source/WebCore/css/SelectorChecker.cpp:
```diff
@@ -1418,6 +1418,34 @@ bool SelectorChecker::matchHasPseudoClass(CheckingContext& checkingContext, cons
+        return false;
+    };
+
+    for (auto& relation : hasCheckingContext.styleRelations) {
+        if (relationNeededForHasInvalidation(relation.type))
+            checkingContext.styleRelations.append(relation);
+    }
```
Ahha. Good find.

anttijk replied:
I suppose this is really a different case (children are not affected, rather the current element is affected by the children/descendants) though using the same bits may be reasonable.

anttijk replied:
Wonder if there is an alternative solution that would add new `MatchElement` types for the case where both additions and removals need to be always checked? Or maybe `IsNegation` enum could become three-valued one to handle that situation in general way.

heycam replied:
Yes, I think that could work too. And actually might be a better solution. But I've run out of time, so I will file a followup for that.

heycam replied:
Filed https://bugs.webkit.org/show_bug.cgi?id=261315.

heycam committed:
Invalidate :has() with nested positional pseudo-class selectors correctly
https://bugs.webkit.org/show_bug.cgi?id=253943
rdar://106768224

Reviewed by Antti Koivisto.

ChildChangeInvalidation currently handles :has() invalidation by
traversing the newly inserted or removed subtree and checking whether
there are any :has() rules whose nested selector could match the elements
in the subtree. This is not sufficient to invalidate rules like:

:has(:first-child)
:has(:nth-child(2n+1))
:has(:nth-child(n2+1 of .foo))

When the parent of the changed subtree has any of these flags:

ChildrenAffectedByFirstChildRules
ChildrenAffectedByLastChildRules
ChildrenAffectedByForwardPositionalRules
ChildrenAffectedByBackwardPositionalRules

we must invalidate some siblings of the subtree, to determine if the
:has() selector might have changed.

The "check the existing sibling" optimization that
ChildChangeInvalidation::invalidateForChangedElement does is not valid
if it's called with a sibling of the changed subtree, so we pass in
a ChangedElementRelation value to control whether to skip it.

* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/child-indexed-pseudo-classes-in-has-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/typed-child-indexed-pseudo-classes-in-has-expected.txt:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::matchHasPseudoClass const):
* Source/WebCore/style/ChildChangeInvalidation.cpp:
(WebCore::Style::ChildChangeInvalidation::invalidateForChangedElement):
(WebCore::Style::ChildChangeInvalidation::invalidateForHasBeforeMutation):
(WebCore::Style::ChildChangeInvalidation::invalidateForHasAfterMutation):
(WebCore::Style::ChildChangeInvalidation::traverseRemovedElements):
(WebCore::Style::ChildChangeInvalidation::traverseAddedElements):
(WebCore::Style::ChildChangeInvalidation::traverseRemainingExistingSiblings):
* Source/WebCore/style/ChildChangeInvalidation.h:

Canonical link: https://commits.webkit.org/267771@main
